ons-spark/spark-concepts/partitions.md

```diff
@@ -0,0 +1,941 @@
+
+```
+````
+
+```plaintext
++-------+-------+
+|part_id|  count|
++-------+-------+
+|     78|  90000|
+|     43|9900000|
+|     49|    900|
+|    106|    100|
+|     89|   9000|
++-------+-------+
+
+Number of partitions: 200
+```
+
+````{tabs}
+```{code-tab} py
+get_partitioning_info(grouped_df)
+```
+
+```{code-tab} r R
+
+get_partitioning_info(grouped_df)
+
+```
+````
+
+```plaintext
++-------+-----+
+|part_id|count|
++-------+-----+
+|     78|    1|
+|     43|    1|
+|     49|    1|
+|    106|    1|
+|     89|    1|
++-------+-----+
+
+Number of partitions: 200
+```
+
+````{tabs}
+```{code-tab} py
+get_partitioning_info(window_df)
+```
+
+```{code-tab} r R
+
+get_partitioning_info(window_df)
+
+```
+````
+
+```plaintext
++-------+-------+
+|part_id|  count|
++-------+-------+
+|      1|5000000|
+|      0|5000000|
++-------+-------+
+
+Number of partitions: 200
+```
+### Discussion
+
+Now that we have all the information on how Spark performed these processes we can compare the three cases.
+
+All three operations involve an intermediate shuffle where the join key/grouped variable/windows are moved to the same partitions. This is called `hashpartitioning()` in the execution plan. As we mentioned previously, these are all wide transformations so the number of partitions of the output DataFrames are the same, 200. Listed below are the differences in processing and output DataFrames for the three operations.
+
+**Join**
+
+There was much more work to be done on the larger partition than the smaller ones. The Spark UI shows there were more records to process on the large partition therefore the task takes much longer. Note the number of records processed in each task matches the number of rows in each join key. The output DataFrame is partitioned by the join key. 
+
+**Groupby**
+
+In the case of the groupby the skewed variable isn't too much of an issue because the aggregation is quite simple and is optimised, therefore it is easy to process each partition whether there is skew or not. Again, the output DataFrame is partitioned by the grouped variable, this time there is one row on each partition.
+
+**Window**
+
+Like the join, there is more work to do on the larger partitions. However, unlike the join, the rows of the output DataFrame have the same partition IDs as the input DataFrame.
+
+### What does this mean in practice?
+
+Highly skewed data is a common issue that causes slow processing with Spark. Above we have seen that skewed data causes skewed partitions when Spark processes wide transformations and can result in bottlenecks, where some tasks take much longer than others therefore not utilising the full potential of parallel processing. An even worse situation is where the skew causes spill, where the large partition cannot fit into it's allocated memory on an executor and overflows temporarily onto disk. We cannot recreate the issue of spill in a local session but they are easy to spot in the Stage details page in the Spark UI.
```

Comment:
You don't need an apostrophe in its here

Reply:
done